### PR TITLE
staticaddr: channel funding with deposits

### PR DESCRIPTION
**Pull request by hieblmi**
_depends on https://github.com/lightninglabs/loop/pull/1073_

This PR introduces a `openchannel` subcommand to static addresses.

It provides the same experience as `lncli openchannel`. 
Exmples:

Open a channel with all available deposits:
```
loop static openchannel --fundmax --sat_per_vbyte XXX --private ...
```

Open a channel with specified local funding amount, coin-selected from available deposits under fee and dust considerations.
```
loop static openchannel --local_amt XXX --sat_per_vbyte YYY --private ...
```

Open a channel from two deposits AAA and BBB while taking their combined value(fundmax) as funding amount and considering fees and dust limit.
```
loop static openchannel --utxo AAA --utxo BBB --fundmax --sat_per_vbyte YYY --private ...
```

Open a channel from two deposits AAA and BBB while taking a specified amount(local_amt) as funding amount and considering fees and dust limit.
```
loop static openchannel --utxo AAA --utxo BBB --local_amt --sat_per_vbyte YYY --private ...
```

TODOs:

- [x] Integration tests

### CODE REVIEW

File: cmd/loop/openchannel.go [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+		},
+	},
+	Action: openChannel,
+}
+
+func openChannel(ctx context.Context, cmd *cli.Command) error {
+	var (
+		args      = cmd.Args()
+		remaining []string
+		ctxb      = context.Background()
```
[the remark this answers is not in the document]
Comment (hieblmi) [reply cmd/loop/openchannel.go]: fixed

File: cmd/loop/openchannel.go [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+		req.Outpoints = outpoints
+	}
+
+	// The fundmax flag is NOT allowed to be combined with local_amt above.
+	// It is allowed to be combined with push_amt, but only if explicitly
+	// set.
+	if cmd.Bool("fundmax") && req.LocalFundingAmount != 0 {
+		return fmt.Errorf("local amount cannot be set if attempting " +
+			"to commit the maximum amount out of the wallet")
+	}
```
[the remark this answers is not in the document]
Comment (hieblmi) [reply cmd/loop/openchannel.go]: fixed

File: cmd/loop/openchannel.go [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+		req.CommitmentType = lnrpc.CommitmentType_SIMPLE_TAPROOT
+	default:
+		return fmt.Errorf("unsupported channel type %v", channelType)
+	}
+
+	wrappedReq := &looprpc.StaticOpenChannelRequest{
+		OpenChannelRequest: req,
+	}
+
+	resp, err := client.StaticOpenChannel(ctxb, wrappedReq)
```
[the remark this answers is not in the document]
Comment (hieblmi) [reply cmd/loop/openchannel.go]: fixed

File: cmd/loop/openchannel.go [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+		RemoteChanReserveSat:       cmd.Uint64("remote_reserve_sats"),
+		Memo:                       cmd.String("memo"),
+	}
+
+	switch {
+	case cmd.IsSet("node_key"):
+		nodePubHex, err := hex.DecodeString(cmd.String("node_key"))
+		if err != nil {
+			return fmt.Errorf("unable to decode node public key: "+
+				"%v", err)
```
[the remark this answers is not in the document]
Comment (hieblmi) [reply cmd/loop/openchannel.go]: fixed

**Comment by hieblmi** [comment]
@gemini-code-assist review

File: staticaddr/deposit/fsm.go [hunk 5]
```diff
@@ -366,6 +387,17 @@ func (f *FSM) DepositStatesV0() fsm.States {
...
+		OpeningChannel: fsm.State{
+			Transitions: fsm.Transitions{
+				fsm.OnError:        Deposited,
+				OnChannelPublished: ChannelPublished,
+				OnRecover:          OpeningChannel,
+			},
+			Action: fsm.NoOpAction,
+		},
+		ChannelPublished: fsm.State{
+			Action: fsm.NoOpAction,
```
[the remark this answers is not in the document]
Comment (hieblmi) [reply staticaddr/deposit/fsm.go]: @gemini-code-assist `f.FinalizeDepositAction` removes the fsm from the set of active fsms. Isn't it needed later once the channel is active?

**Comment by hieblmi** [comment]
@gemini-code-assist review

File: staticaddr/openchannel/manager.go [hunk 6]
```diff
@@ -0,0 +1,659 @@
...
+		deposits, err = m.cfg.DepositManager.GetActiveDepositsInState(
+			deposit.Deposited,
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		if req.LocalFundingAmount != 0 {
+			deposits, err = staticutil.SelectDeposits(
+				deposits, req.LocalFundingAmount,
```
Comment (sputn1ck) [inline staticaddr/openchannel/manager.go:334]: SelectDeposits isn’t fee-aware. We lock deposits based on local_amt + dust, then may fail once fees are applied. Worth a fee-aware selection or pre-check before OnOpeningChannel.

File: staticaddr/openchannel/manager.go [hunk 7]
```diff
@@ -0,0 +1,659 @@
...
+		}
+	}
+
+	// We need to transition the deposits to the opening channel state
+	// before we start the channel open process. This is important to
+	// ensure that the deposits are not used for other purposes while we
+	// are opening the channel.
+	err = m.cfg.DepositManager.TransitionDeposits(
+		ctx, deposits, deposit.OnOpeningChannel, deposit.OpeningChannel,
+	)
```
Comment (sputn1ck) [inline staticaddr/openchannel/manager.go:365]: Once deposits move to OpeningChannel there’s no recovery path here. On restart, these can stay locked indefinitely unless something else reconciles. Should we reconcile with lnd pending chans / on-chain spends?

File: staticaddr/openchannel/manager.go [hunk 8]
```diff
@@ -0,0 +1,659 @@
...
+		chanCommitmentType = lnrpc.CommitmentType_ANCHORS
+
+	case lnrpc.CommitmentType_SIMPLE_TAPROOT:
+		chanCommitmentType = lnrpc.CommitmentType_SIMPLE_TAPROOT
+	}
+
+	if req.SatPerVbyte == 0 {
+		// Get the fee rate for the withdrawal sweep.
+		feeRate, err = m.cfg.WalletKit.EstimateFeeRate(
+			ctx, defaultConfTarget,
```
Comment (sputn1ck) [inline staticaddr/openchannel/manager.go]: We estimate fee rate here, but CreateFinalizedWithdrawalTx estimates again when sat_per_vbyte is unset. That can skew funding amt vs PSBT build. Can we thread a single fee rate through?

File: staticaddr/openchannel/manager.go [hunk 9]
```diff
@@ -0,0 +1,659 @@
...
+
+	openChanRequest := &lnrpc.OpenChannelRequest{
+		NodePubkey:                 req.NodePubkey,
+		LocalFundingAmount:         int64(chanFundingAmt),
+		PushSat:                    req.PushSat,
+		Private:                    req.Private,
+		MinHtlcMsat:                req.MinHtlcMsat,
+		RemoteCsvDelay:             req.RemoteCsvDelay,
+		MinConfs:                   defaultUtxoMinConf,
+		SpendUnconfirmed:           req.SpendUnconfirmed,
```
Comment (sputn1ck) [inline staticaddr/openchannel/manager.go]: MinConfs is hard-coded to 1. If caller sets MinConfs !=1, we silently ignore it. Should we validate/reject unsupported MinConfs before locking deposits?

File: staticaddr/openchannel/manager.go [hunk 10]
```diff
@@ -0,0 +1,659 @@
...
+		Memo:                       req.Memo,
+	}
+
+	chanTxHash, err := m.openChannelPsbt(
+		ctx, openChanRequest, deposits, int64(req.SatPerVbyte),
+	)
+	if err != nil {
+		log.Infof("error opening channel: %v", err)
+		err2 := m.cfg.DepositManager.TransitionDeposits(
+			ctx, deposits, fsm.OnError, deposit.Deposited,
```
Comment (sputn1ck) [inline staticaddr/openchannel/manager.go]: Any error after PSBT finalize currently transitions deposits back to Deposited. If finalize already handed the tx to lnd (NoPublish=false), we can reuse spent deposits. Can we gate rollback once finalize is sent or persist txid/reconcile?

File: staticaddr/openchannel/managert_test.go [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+	"github.com/lightningnetwork/lnd/input"
+	"github.com/lightningnetwork/lnd/lnrpc"
+	"github.com/lightningnetwork/lnd/lnwallet"
+	"github.com/lightningnetwork/lnd/lnwallet/chainfee"
+	"github.com/stretchr/testify/require"
+)
+
+// TestCalculateFundingTxValues tests the calculateFundingTxValaues function
+// with various scenarios.
+func TestCalculateFundingTxValues(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (hieblmi) [reply staticaddr/openchannel/managert_test.go]: fixed

File: staticaddr/openchannel/manager.go (hunk 6 above)
Comment (hieblmi) [reply staticaddr/openchannel/manager.go:334]: I made `SelectDeposits` fee aware.

File: staticaddr/openchannel/manager.go (hunk 7 above)
Comment (hieblmi) [reply staticaddr/openchannel/manager.go:365]: Good point. This needs reconciliation on startup to recover deposits stuck in OpeningChannel state. Will address in a follow-up PR that checks lnd pending channels and on-chain spends to transition deposits back to Deposited or forward to ChannelPublished.

File: staticaddr/openchannel/manager.go (hunk 8 above)
Comment (hieblmi) [reply staticaddr/openchannel/manager.go]: Fixed

File: staticaddr/openchannel/manager.go (hunk 10 above)
Comment (hieblmi) [reply staticaddr/openchannel/manager.go]: Gated rollback after we send out the finalize.

File: staticaddr/openchannel/manager.go (hunk 9 above)
Comment (hieblmi) [reply staticaddr/openchannel/manager.go]: done

**Comment by litbot-9000** [comment]
@sputn1ck: review reminder
@hieblmi, remember to re-request review from reviewers when ready

**Comment by hieblmi** [comment]
Hi @sputn1ck, I addressed some more findings, namely:
- Added crash recovery for OpeningChannel deposits.
- Upgraded permissions to swap:execute for both StaticOpenChannel and StaticAddressLoopIn.
- Aligned channel-open fee/value calculation paths by threading commitment type into tx construction.
- Hardcoded forwarded SpendUnconfirmed to false.
- Added unit tests for openchannel recovery logic.

File: staticaddr/openchannel/manager.go [hunk 12]
```diff
@@ -0,0 +1,806 @@
...
+			// If the remote peer canceled on us, the reservation
+			// has already been deleted. We don't need to try to
+			// remove it again, this would just produce another
+			// error.
+			cancelErr := chanfunding.ErrRemoteCanceled.Error()
+			if err != nil && strings.Contains(
+				err.Error(), cancelErr,
+			) {
+
+				shimPending = false
```
Comment (starius) [inline staticaddr/openchannel/manager.go:587]: Data race in shimPending variable. It is accessed from multiple goroutines without synchronization.

File: staticaddr/openchannel/manager.go [hunk 13]
```diff
@@ -0,0 +1,806 @@
...
+			// interrupts now, we don't need to clean up anything.
+			shimPending = false
+
+			hash, err := chainhash.NewHash(
+				update.ChanPending.Txid,
+			)
+			if err != nil {
+				log.Infof("Error creating hash for channel "+
+					"open tx: %v", err)
+			}
```
Comment (starius) [inline staticaddr/openchannel/manager.go:721]: I think we should `return error` in this `if`. We use the `hash` variable below as if it is valid, but it could be `nil` and we just log the error and crash later.

File: staticaddr/openchannel/manager.go [hunk 14]
```diff
@@ -0,0 +1,806 @@
...
+			log.Infof("Please monitor the channel from lnd")
+
+			err = m.cfg.DepositManager.TransitionDeposits(
+				ctx, deposits, deposit.OnChannelPublished,
+				deposit.ChannelPublished,
+			)
+			if err != nil {
+				log.Errorf("error transitioning deposits to "+
+					"ChannelPublished: %v", err)
+			}
```
Comment (starius) [inline staticaddr/openchannel/manager.go:734]: `TransitionDeposits` errors are logged but the function still returns success, leaving deposits stuck in `OpeningChannel` while the channel is actually pending. This should be returned to the caller or retried.

File: staticaddr/deposit/fsm.go [hunk 15]
```diff
@@ -366,6 +387,17 @@ func (f *FSM) DepositStatesV0() fsm.States {
...
+			Transitions: fsm.Transitions{
+				fsm.OnError:        Deposited,
+				OnChannelPublished: ChannelPublished,
+				OnRecover:          OpeningChannel,
+			},
+			Action: fsm.NoOpAction,
+		},
+		ChannelPublished: fsm.State{
+			Action: f.FinalizeDepositAction,
+		},
```
Comment (starius) [inline staticaddr/deposit/fsm.go:412]: These two states lack OnExpiry transitions from them. OnExpiry may be delivered to each of them.

1. OpeningChannel has Action: fsm.NoOpAction. This means it stops current FSM execution and unlocks the FSM mutex, letting potential SendEvent calls sneak in.
2. FinalizeDepositAction also returns fsm.NoOp and triggers further things via a channel (f.finalizedDepositChan) so there is still a window when SendEvent can sneak in.

`handleBlockNotification` fires OnExpiry on every new block once the deposit is expired (staticaddr/deposit/fsm.go), regardless of state.

So there is a possibility that OnExpiry is delivered to each of these new states. We have to add state self-transitions to each of them.

File: staticaddr/openchannel/manager.go [hunk 16]
```diff
@@ -0,0 +1,806 @@
...
+	}
+
+	if len(openingDeposits) == 0 {
+		return nil
+	}
+
+	log.Infof("Recovering %d deposits in OpeningChannel state",
+		len(openingDeposits))
+
+	utxos, err := m.cfg.WalletKit.ListUnspent(
```
Comment (starius) [inline staticaddr/openchannel/manager.go:180]: What if the deposit is spent, but not confirmed? Shouldn't we still track it? Maybe it is evicted from mempool. Or even confirmed first and then reorged.

File: staticaddr/openchannel/manager_test.go [hunk 17]
```diff
@@ -0,0 +1,314 @@
...
+			if tc.expectedErrSubstr == "" {
+				require.NoError(t, err)
+				require.Equal(t, tc.expectedType, commitmentType)
+				return
+			}
+
+			require.ErrorContains(t, err, tc.expectedErrSubstr)
+		})
+	}
+}
```
Comment (starius) [inline staticaddr/openchannel/manager_test.go:597]: I propose to add test coverage for edge cases:
 - reorg
 - daemon restart during channel opening
 - eviction from mempool
 - rejection from mempool
 - stream errors
 - PSBT finalize then stream abort
 - duplicate outpoints in request

File: staticaddr/openchannel/manager.go [hunk 18]
```diff
@@ -0,0 +1,806 @@
...
+		newOpenChannelRequestChan: make(chan newOpenChannelRequest),
+		errChan:                   make(chan error),
+	}
+
+	return m
+}
+
+// Run runs the open channel manager.
+func (m *Manager) Run(ctx context.Context) error {
+	err := m.recoverOpeningChannelDeposits(ctx)
```
Comment (starius) [inline staticaddr/openchannel/manager.go:125]: Should we recover deposits during daemon running as well, not only during startup time?

If the PSBT finalize succeeds but the stream fails before `ChanPending`, deposits remain in `OpeningChannel` until restart.

File: staticaddr/openchannel/manager.go [hunk 19]
```diff
@@ -0,0 +1,806 @@
...
+			req.SatPerVbyte * 1000,
+		).FeePerKWeight()
+	}
+
+	// There are three ways in which we select deposits to open a channel
+	// with. 1.) The user manually selects the deposits. 2.) The user only
+	// selects a local channel amount in which case we coin-select deposits
+	// to cover for it. 3.) The user selects the fundmax flag, in which case
+	// we select all deposits to fund the channel.
+	if len(req.Outpoints) > 0 {
```
Comment (starius) [inline staticaddr/openchannel/manager.go:294]: What if `req.Outpoints` has duplicates? This leads to fee miscalculation and invalid PSBT building (having the same input twice). I propose to return an error early if we detect duplicates in `req.Outpoints` and add test coverage.

File: staticaddr/deposit/manager.go
Comment (starius) [inline staticaddr/deposit/manager.go]: dead code

File: looprpc/client.proto [hunk 20]
```diff
@@ -206,13 +207,33 @@ service SwapClient {
...
+    // Wrap lnd's request so Loop can extend this RPC with Loop-specific fields
+    // without diverging from the upstream API surface.
+    lnrpc.OpenChannelRequest open_channel_request = 1;
+}
+
+message StaticOpenChannelResponse {
+    /*
+    The transaction hash of the channel opening transaction.
+    */
+    string channel_open_tx_hash = 1;
```
Comment (starius) [inline looprpc/client.proto]: Should we also add the output index (always 0 in the current implementation)? Just in case we implement batch open or something in the future. The code which uses this thing shouldn't guess what is the index, it should take it from the API.

Comment (starius) [reply looprpc/client.proto]: Alternatively change `channel_open_tx_hash` to `channel_open_outpoint` and put "txid:index" there.

File: staticaddr/deposit/manager.go
Comment (hieblmi) [reply staticaddr/deposit/manager.go]: fixed

File: staticaddr/openchannel/manager.go (hunk 12 above)
Comment (hieblmi) [reply staticaddr/openchannel/manager.go:587]: Added a mutex to gate access.

File: staticaddr/openchannel/manager.go (hunk 13 above)
Comment (hieblmi) [reply staticaddr/openchannel/manager.go:721]: Good catch, fixed.

File: staticaddr/openchannel/manager.go (hunk 14 above)
Comment (hieblmi) [reply staticaddr/openchannel/manager.go:734]: The current approach is to just log the transition error and return the `chanOutpoint` since the transition error might be a recurring error that isn't solved without intervention. 
So the approach is to reconcile these `OpeningChannel` states after a restart in `recoverOpeningChannelDeposits`.

File: staticaddr/deposit/fsm.go (hunk 15 above)
Comment (hieblmi) [reply staticaddr/deposit/fsm.go:412]: That is a great catch, thank you!
Fixed it.

File: staticaddr/openchannel/manager.go (hunk 18 above)
Comment (hieblmi) [reply staticaddr/openchannel/manager.go]: I've fixed that in commit 15bfd345d1a2050ea27d50f39499c97dacd9d60d

File: staticaddr/openchannel/manager.go (hunk 19 above)
Comment (hieblmi) [reply staticaddr/openchannel/manager.go:294]: added logic to detect dupes and return in that case.

File: looprpc/client.proto (hunk 20 above)
Comment (hieblmi) [reply looprpc/client.proto]: I like the second suggestion, added.

File: staticaddr/openchannel/manager.go (hunk 18 above)
Comment (hieblmi) [reply staticaddr/openchannel/manager.go:125]: Fixed. We are checking if there's an error but the PSBT was finalized and sent, then we recover remaining deposits.

File: staticaddr/openchannel/manager_test.go (hunk 17 above)
Comment (hieblmi) [reply staticaddr/openchannel/manager_test.go:597]: Added!

File: staticaddr/openchannel/manager.go (hunk 14 above)
Comment (starius) [reply staticaddr/openchannel/manager.go:734]: What do we do with TransitionDeposits errors? Can we retry TransitionDeposits several times and if it still fails, run recoverOpeningChannelDeposits? And if it also fails, return hard error.
Which errors TransitionDeposits can return? Don't we hide anything important when skipping the errors?

File: staticaddr/openchannel/manager.go [hunk 21]
```diff
@@ -0,0 +1,837 @@
...
+		CommitmentType:             chanCommitmentType,
+		ZeroConf:                   req.ZeroConf,
+		ScidAlias:                  req.ScidAlias,
+		BaseFee:                    req.BaseFee,
+		FeeRate:                    req.FeeRate,
+		UseBaseFee:                 req.UseBaseFee,
+		UseFeeRate:                 req.UseFeeRate,
+		RemoteChanReserveSat:       req.RemoteChanReserveSat,
+		Memo:                       req.Memo,
+	}
```
Comment (starius) [inline staticaddr/openchannel/manager.go]: Most of the fields are copied from `req`.
Could we just clone `req` and then adjust few fields (SpendUnconfirmed, MinConfs, etc).

Or even better accept `req` as-is and apply checks on its fields. E.g. demand the correct value of LocalFundingAmount, MinConfs, SpendUnconfirmed) in the input `req`.

Note that some fields are silently ignored or skipped when passing to LND:
  - target_conf: ignored (not used for fee selection, not rejected).
  - sat_per_byte (deprecated): ignored (not used, not rejected).
  - node_pubkey_string (deprecated): ignored by this path (only node_pubkey bytes are used).
  - funding_shim: ignored/overwritten (manager always builds its own PSBT shim).
  - sat_per_vbyte: used locally for withdrawal/funding fee math, but not forwarded to lnd OpenChannel request.
  - fund_max, outpoints: consumed by Loop’s own deposit-selection logic, not forwarded.
  - min_confs: normalized to 1 for PSBT flow (0 accepted as default/unset).

Shouldn't we pass fund_max to LND? It affects channel size selection.

Comment (hieblmi) [reply staticaddr/openchannel/manager.go]: I think your second option is good, I will apply `Or even better accept req as-is and apply checks on its fields`.
However, `fundmax` cannot be passed to lnd in psbt mode as that isn't implemented yet.

**Comment by hieblmi** [comment]
@starius addressed recent comments in commit e0934b6a2a3cc4dead1dbaa059f5af47d8e65ced

File: staticaddr/openchannel/manager.go [hunk 22]
```diff
@@ -0,0 +1,849 @@
...
+					"after failed channel open: %v",
+					err2)
+			}
+		}
+
+		return nil, err
+	}
+
+	return chanOutpoint, nil
+}
```
Comment (starius) [inline staticaddr/openchannel/manager.go]: Can we reduce indentation in this code?

```go
chanOutpoint, err := m.openChannelPsbt(ctx, reqClone, deposits, feeRate)
if err == nil {
  return chanOutpoint, nil
}

log.Infof("error opening channel: %v", err)
...

```

File: staticaddr/openchannel/manager.go [hunk 23]
```diff
@@ -447,6 +447,7 @@ func (m *Manager) openChannelPsbt(ctx context.Context,
 		psbtFinalized bool
 		basePsbtBytes []byte
 		quit          = make(chan struct{})
+		quitErr       = make(chan error, 1)
```
Comment (starius) [inline staticaddr/openchannel/manager.go]: Can we merge these two channels together?

Comment (hieblmi) [reply staticaddr/openchannel/manager.go]: We can remove the chan and just assign the error instead.

File: staticaddr/openchannel/manager.go (hunk 22 above)
Comment (hieblmi) [reply staticaddr/openchannel/manager.go]: done
